cypress/support/commands.js

```
@@ -26,37 +26,41 @@
 // Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
 //
 
+/// <reference types="cypress" />
+
 Cypress.Commands.add('getByDataCy', (selector) => {
-  cy.get(`[data-cy^="${selector}"]`);
+  return cy.get(`[data-cy^="${selector}"]`);
 });
 
-Cypress.Commands.add('register', (email = 'riot@qa.team', username = 'riot', password = '12345Qwert!') => {
-  cy.request('POST', '/api/users', {
-    user: {
-      email,
-      username,
-      password
+Cypress.Commands.add(
+  'register',
+  (email = 'riot@qa.team', username = 'riot', password = '12345Qwert!') => {
+    cy.request('POST', '/api/users', {
+      user: { email, username, password },
+    });
+  }
+);
+
+Cypress.Commands.add('login', (email, username, password) => {
+  cy.request({
+    method: 'POST',
+    url: '/api/users/login',
+    body: { user: { email, password } },
+    failOnStatusCode: false,
+  }).then((resp) => {
+    if (resp.status === 200 && resp.body?.user?.token) {
+      window.localStorage.setItem('jwt', resp.body.user.token);
+      cy.setCookie('auth', resp.body.user.token);
+      cy.visit('/');
+    } else {
+      cy.visit('/login');
+      cy.get('[data-cy="login-email-input"]').type(email);
+      cy.get('[data-cy="login-password-input"]').type(password);
+      cy.get('[data-cy="login-submit-btn"]').click();
     }
   });
```

The `cy.request` uses `failOnStatusCode: false` and falls back to a UI login flow on non-200. Falling back to UI interactions in the helper may mask backend issues and make tests slower/flaky. Prefer failing fast (so tests surface why programmatic login failed) or return the request and let the test decide the next action. If you keep the fallback, ensure the UI selectors used are stable and that the command returns a chainable that resolves after the UI login completes.